File: services/oracle/oracle_service/sage_client.py
from __future__ import annotations

"""Small Sage HTTP client used for Oracle's read-only broker calls."""

import json
from urllib.parse import quote

import httpx

SAGE_REQUEST_TIMEOUT_SECONDS = 20.0


class SageClient:
    def __init__(self, base_url: str, service_token: str) -> None:
        self._base_url = base_url.rstrip("/")
        self._headers = {
            "Authorization": f"Bearer {service_token}",
            "Content-Type": "application/json"
        }

    async def _request_json(self, path: str, method: str = "GET", body: object | None = None) -> tuple[bool, int, object | None]:
        try:
            async with httpx.AsyncClient(timeout=SAGE_REQUEST_TIMEOUT_SECONDS) as client:
                response = await client.request(
                    method,
                    f"{self._base_url}{path}",
                    headers=self._headers,
                    json=body
                )
        except httpx.HTTPError as error:
            return False, 0, {
                "error": str(error),
                "type": error.__class__.__name__
            }

        payload: object | None
        text = response.text
        if not text:
            payload = None
        else:
            try:
                payload = response.json()
            except json.JSONDecodeError:
                # Preserve plain-text upstream failures instead of dropping the response body.
                payload = {"raw": text}

        return response.is_success, response.status_code, payload

    async def get_setting(self, key: str) -> object | None:
        _ok, _status, payload = await self._request_json(
            f"/api/internal/vault/settings/{quote(key, safe='')}"
        )
        return payload

    async def get_secret(self, key: str) -> object | None:
        _ok, _status, payload = await self._request_json(
            f"/api/internal/vault/secrets/{quote(key, safe='')}"
        )
        return payload

    async def get_bootstrap_status(self) -> object:
        ok, status, payload = await self._request_json("/api/internal/warden/bootstrap")
        if not ok:
            error = payload["error"] if isinstance(payload, dict) and "error" in payload else f"Sage bootstrap request failed with {status}"
            raise RuntimeError(str(error))
        return payload

File: services/oracle/lib/sageClient.mjs
const withBearer = (token) => ({
  "Authorization": `Bearer ${token}`,
  "Content-Type": "application/json"
});

const json = async (response) => {
  const text = await response.text();
  if (!text) {
    return null;
  }
  try {
    return JSON.parse(text);
  } catch {
    return {raw: text};
  }
};

const requestJson = async (baseUrl, path, headers, options = {}) => {
  const response = await fetch(`${baseUrl}${path}`, {
    method: options.method || "GET",
    headers: {
      ...headers,
      ...(options.headers || {})
    },
    body: options.body == null ? undefined : JSON.stringify(options.body)
  });
  const payload = await json(response);
  return {
    ok: response.ok,
    status: response.status,
    payload
  };
};

export const createSageClient = (config) => {
  const headers = withBearer(config.serviceToken);
  const baseUrl = config.sageBaseUrl;

  return {
    async getSetting(key) {
      const response = await requestJson(baseUrl, `/api/internal/vault/settings/${encodeURIComponent(key)}`, headers);
      return response.payload;
    },
    async getSecret(key) {
      const response = await requestJson(baseUrl, `/api/internal/vault/secrets/${encodeURIComponent(key)}`, headers);
      return response.payload;
    },
    async getBootstrapStatus() {
      const response = await requestJson(baseUrl, "/api/internal/warden/bootstrap", headers);
      if (!response.ok) {
        throw new Error(response.payload?.error || `Sage bootstrap request failed with ${response.status}`);
      }
      return response.payload;
    }
  };
};

export default createSageClient;
